tests/integration/security/authorization_test.go

```diff
@@ -1237,3 +1243,133 @@ func TestAuthorization_Audit(t *testing.T) {
+	if err != nil {
+		ctx.Fatalf("Failed to read config map: %v", err)
+	}
+
+	// Create two extension providers using the same ext-authz service, one for HTTP and the other for gRPC ext-authz API.
+	service := fmt.Sprintf("%s/ext-authz", ns)
+	httpProvider := &v1alpha1.MeshConfig_ExtensionProvider{
+		Name: "ext-authz-http",
+		Provider: &v1alpha1.MeshConfig_ExtensionProvider_EnvoyExtAuthzHttp{
+			EnvoyExtAuthzHttp: &v1alpha1.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationHttpProvider{
+				Service:               service,
+				Port:                  8000,
+				PathPrefix:            "/check",
+				IncludeHeadersInCheck: []string{"x-ext-authz"},
+			},
+		},
+	}
+	grpcProvider := &v1alpha1.MeshConfig_ExtensionProvider{
+		Name: "ext-authz-grpc",
+		Provider: &v1alpha1.MeshConfig_ExtensionProvider_EnvoyExtAuthzGrpc{
+			EnvoyExtAuthzGrpc: &v1alpha1.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationGrpcProvider{
+				Service: service,
+				Port:    9000,
+			},
+		},
+	}
+
+	// Update the mesh config.
+	meshConfig.ExtensionProviders = []*v1alpha1.MeshConfig_ExtensionProvider{httpProvider, grpcProvider}
+	configMap.Data["mesh"], err = protomarshal.ToYAML(meshConfig)
+	if err != nil {
+		ctx.Fatalf("Failed to marshal mesh config: %v", err)
+	}
+	if _, err := ctx.Clusters().Default().CoreV1().ConfigMaps(systemNs.Name()).Update(
+		context.Background(), configMap, v1.UpdateOptions{}); err != nil {
+		ctx.Fatalf("Failed to update config map: %v", err)
+	}
+
+	// Restart istiod to make sure the change to istiod is ready for the following tests.
+	err = ctx.Clusters().Default().CoreV1().Pods(systemNs.Name()).DeleteCollection(
+		context.Background(), kube.DeleteOptionsForeground(), v1.ListOptions{LabelSelector: "app=istiod"})
+	if err != nil {
+		ctx.Fatalf("Failed to restart istiod after mesh config update: %v", err)
+	}
+	if _, err := kube.WaitUntilPodsAreReady(
+		kube.NewPodFetch(ctx.Clusters().Default(), systemNs.Name(), "app=istiod")); err != nil {
+		ctx.Fatalf("Wait for istiod restart failed: %v", err)
+	}
+}
+
+// TestAuthorization_Custom tests that the CUSTOM action with the sample ext_authz server.
+func TestAuthorization_Custom(t *testing.T) {
+	framework.NewTest(t).
+		Features("security.authorization.custom").
+		Run(func(ctx framework.TestContext) {
+			nsExtServer := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "v1beta1-custom-ext-server",
+				Inject: true,
+			})
+			ns := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "v1beta1-custom",
+				Inject: true,
+			})
+			args := map[string]string{"Namespace": ns.Name()}
+			applyYAML := func(filename string, ns namespace.Instance) []string {
+				policy := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, filename))
+				ctx.Config().ApplyYAMLOrFail(t, ns.Name(), policy...)
+				return policy
+			}
+
+			// Configure mesh config to add the definition of the ext-authz server.
+			configureExtensionProviders(t, ctx, nsExtServer.Name())
```

Comment:
Does ext provider config in meshconfig need clean up post test run?

Reply:
no need, it can just be kept there.